### DISCUSSION TITLE
Defining our goals and desired outcomes

### DISCUSSION THREAD

**Discussion by byrnereese** [mkdocs/mkdocs#3670] [General]
First, let me just say that I am honestly honored to have been invited to be a part of this group and initiative. In my day job, I am a Product Manager who specializes in building Developer-oriented products. We use mkdocs at my company, and I am a plugin and theme contributor to the mkdocs community. One way I would like to contribute to this initiative is to help bring a greater level of organization and focus to our efforts so that we can be successful at achieving our goals.

We will be entertaining many different ideas as we explore the ways we want to "bring mkdocs to the next level." To help us better collaborate, I thought it would be helpful for us to share our goals and what we hope to achieve by participating more deeply in this community. 

If we can get greater alignment around our goals and the outcomes we are trying to achieve, we can foster a more supportive environment, more easily avoid conflict with one another, and ultimately ship a product that brings real value to our users, and dare I say, the world. 

So I would like to invite everyone to share: what do you want to achieve for yourself and for the mkdocs project?

**Comment by squidfunk** [comment]
IMHO one of the most important goals is to show that MkDocs is an ecosystem of projects. This is what makes it so powerful. This would entail a rewrite of the documentation to better showcase what can be done with MkDocs, and how the many available plugins and extensions can work together!

Additionally, writing plugins and hooks already is super simple, but we should make it more apparent in the documentation how simple it is. Additionally, giving plugin authors utilities for repetitive things that need to be done in plugins (e.g. filter pages), would make plugins even easier to write.

In short: empowering the ecosystem.

**Reply by EddyLuten** [reply]
> Additionally, giving plugin authors utilities for repetitive things that need to be done in plugins (e.g. filter pages), would make plugins even easier to write.

Yes, plus, since there was talk about potentially making the markdown parser swappable, exposing a layer of abstraction would also be helpful. For example, I currently invoke the Python-Markdown parser manually to get a copy of the TOC for one of my plugins during the `on_files` step. I also tap into the `utils/meta.py` file to extract the meta section from each page during the same step. Some of these operations are either hacky or wasteful and I'd love to have more built-in abstractions for things like these that I can be sure are the same ones used by MkDocs itself.

**Reply by fralau** [reply]
> IMHO one of the most important goals is to show that MkDocs is an ecosystem of projects. This is what makes it so powerful. This would entail a rewrite of the documentation to better showcase what can be done with MkDocs, and how the many available plugins and extensions can work together!

A common doc? I like that idea.

**Comment by pawamoy** [comment]
Thanks @byrnereese, your experience as a product manager is super welcome :slightly_smiling_face: I'm glad you're part of this community :smile: 

> what do you want to achieve for yourself and for the mkdocs project?

For my part, I simply want to continue working full-time on the things I love. mkdocstrings is one of those things. Since mkdocstrings entirely depends on MkDocs, I want MkDocs to thrive. Even without mkdocstrings, I'd have become a user of MkDocs by now with all the great plugins we have in the ecosystem, so as a user I also want MkDocs and its ecosystem to thrive.

How to achieve this? By empowering the ecosystem as @squidfunk puts it. By being easy to use, yet very flexible. By providing clear solutions for common use-cases, without drowning users in endless possibilities. By supporting all actors in the ecosystem some way or another (giving back time, visibility, or funding), particularly maintainers of MkDocs' own dependencies. Python-Markdown, which @waylan maintains, and Jinja from the @pallets organization are essential parts of MkDocs, and we should show them some support.

**Comment by d0ugal** [comment]
Thanks, I think this is a useful exercise!

> what do you want to achieve for yourself and for the mkdocs project?

For me, nothing in particular. I'm an old timer I suppose. I done a lot of work on MkDocs back in the day but I've been inactive for years. I burned out a bit on the project. I reached a point where I really wanted to do a v2 and fix/change a bunch of decisions but never had the time to do it. I then became a dad etc. I do still ponder the idea as a mental exercise sometimes!

However, I still smile when I regularly see projects using MkDocs. I care about two things in terms of the project direction.

- ease of use, particularly for new users
- extensibility, particularly for bigger more advanced users.


Or, to put it another way it should be really easy to start using MkDocs and then you should have no reason to stop because the ecosystem has you covered for advanced/varied uses.

But, being an inactive old timer, my views don't matter so much as I'm not likely to find time to commit.

**Comment by waylan** [comment]
Like @d0ugal, I burned out on the project after he handed the torch off to me. I also, would like to change some decisions I made early on. When I started, we were not yet at version 1.0 and IMO we put too much emphasis on backward compatibility. In those early days we should have been more focused on making breaking changes for a better future. But, now that we are past version 1.0 I would prefer to see the existing ecosystem continue to be supported without breaking things. Yes, I realize those two things are in opposition to each other. Maybe a cutoff needs to be made and then a new major version could make breaking changes. That way all of us who are satisfied with the current state of things can continue to use our existing sites unaltered on classic MkDocs without holding back the future of MkDocs.

As an aside, I learned a lot from maintaining MkDocs. If I was to start a new static site generator from scratch, I would do some things very different, but that would not look much like MkDocs and would be a different project entirely. I don't expect to ever actually build it and would like to continue using MkDocs without the need to update my sites for future changes.

**Reply by byrnereese** [reply]
> As an aside, I learned a lot from maintaining MkDocs. If I was to start a new static site generator from scratch, I would do some things very different, but that would not look much like MkDocs and would be a different project entirely. I don't expect to ever actually build it and would like to continue using MkDocs without the need to update my sites for future changes.

@waylan I would be super curious to hear from you more about this. Can you create another conversation on the topic of, "if we could do it all over again?" A complete rebuild of mkdocs might not seem realistic at this point, but it would be useful to know the ways in members of the project feel the product could be improved at a foundational level.

**Reply by waylan** [reply]
@byrnereese I recorded my ideas [here](https://gist.github.com/waylan/1e247f1c331c0abc2d41679f00522813). Although, that is probably not what you were expecting.

**Comment by byrnereese** [comment]
Let me ask a follow up question to the above, and invite people who have already responded to respond again. 

What does success look like for this initiative? What metrics might you look at to know we have achieved our goals?

**Reply by ojacques** [reply]
Good docs for "mkdocs core"? I never contributed to the doc - please apologize. But I feel the mkdocs docs does not do it justice. Am I alone?

**Reply by byrnereese** [reply]
You are not. As a documentation tool, its documentation should better model best-in-class. 

Material Mkdocs to me is our benchmark. Easy to use. Easy to read. And super helpful. (cc: @squidfunk)

**Comment by byrnereese** [comment]
I am also curious to know who here has a commercial stake in the success of mkdocs. For example, it feels obvious to me that the more successful mkdocs is, the more @squidfunk is likely to benefit as it would certainly drive business to mkdocs-material. Is there anyone else here who derives revenue as a mkdocs expert, service provider, or consultant?

(no judgement here, just curious...)

**Reply by pawamoy** [reply]
Yes, me! I follow a sponsorware strategy too ([documented here](https://pawamoy.github.io/insiders/)), where my sponsors get access to private features and projects. I quit my job to work full-time on my open source projects. mkdocstrings is my most popular project, so I believe people/companies generally sponsor me for my work on it. Therefore I do rely on MkDocs being successful :slightly_smiling_face:

**Reply by squidfunk** [reply]
All businesses using MkDocs. I am seriously baffled how many enterprises around the world use (Material for) MkDocs, and there are __a lot__. (Material for) MkDocs has become a very popular solution for creating documentation, but also internal knowledge bases. Some of those enterprises and many individuals sponsor our (I'm also building a team) work, because we closely listen to them and create things that they need, but most of them don't and just enjoy the fruits of it, and that's fine.

**Reply by fralau** [reply]
@squidfunk I agree that MkDocs (and Material) are used by a lot of companies. 

I don't do that but I think it's beautiful, that you can draw income from it.

**Reply by ojacques** [reply]
I am not making money out of mkdocs per say. Yet, I use it at work, and convince many customers to do tech docs using mkdocs. In that sense, I believe it is a commercial use.

**Comment by ojacques** [comment]
> @byrnereese I recorded my ideas [here](https://gist.github.com/waylan/1e247f1c331c0abc2d41679f00522813). Although, that is probably not what you were expecting.

New interesting concepts there. I came to mkdocs because it is (almost) as readable as pure markdown when browsing repos in GitHub/GitLab as after having gone through mkdocs. To me, staying very close to the file structure and pure markdown is what made mkdocs compelling to me (compared to other static site generators like hugo). But maybe I misunderstood the concepts in the proposal.
